db/models/models.js

```diff
@@ -7,11 +9,30 @@ exports.fetchAllTopics = () => {
   });
 };
 
-exports.fetchArticles = () => {
-  return db.query(`SELECT articles.* , COUNT(comments.article_id) AS comment_count FROM articles LEFT JOIN comments on comments.article_id = articles.article_id GROUP BY articles.article_id ORDER BY articles.created_at DESC;`).then((results) => {
-    return results.rows;
-  });
-};
+exports.fetchArticles = async (sortby = 'created_at', order = 'DESC', topics) => {
+  if (order !== 'DESC' && order !== 'ASC'){
+    return Promise.reject({ status: 404, msg: 'order not valid - must be ASC or DESC' });
+    }
+  if (sortby !== 'created_at' && sortby !== 'title' && sortby !== 'article_id' && sortby !== 'topic' && sortby !== 'author' && sortby !== 'body' && sortby !== 'votes'){
+    return Promise.reject({ status: 404, msg: 'sort by catagory does not exist' });
+  }
+  if (topics === undefined){
+    return db.query(`SELECT articles.* , COUNT(comments.article_id) AS comment_count FROM articles LEFT JOIN comments on comments.article_id = articles.article_id GROUP BY articles.article_id ORDER BY articles.${sortby} ${order};`).then((results) => {
+      return results.rows;})
+  } 
+  if(topics.length > 0){
+   const dbOutput = await db.query(
+      `SELECT * FROM articles WHERE articles.topic = $1;`,[topics]
+    );
+    if (dbOutput.rows.length === 0) {
+      return Promise.reject({ status: 404, msg: 'topic not found' });
+    } else {
+      return db.query(`SELECT articles.* , COUNT(comments.article_id) AS comment_count FROM articles LEFT JOIN comments on comments.article_id = articles.article_id  WHERE articles.topic = $1 GROUP BY articles.article_id ORDER BY articles.${sortby} ${order};`,[topics]).then((results) => {
+        return results.rows;})
+    }
+
```

Comment:
rather than doing two separate db.queries depending on if topics is defined or not, I'd like to see an optional WHERE clause put in here https://notes.northcoders.com/courses/js-back-end/complex-queries

It dries up your code. What this would mean is building a query string up conditionally and then right at the end using it in a db.query

I now think the check on line 25 would cause issues with my test suggestion of checking a valid topic with no articles associated with it isn't throwing an error - one to consider, perhaps psql will throw a more useful error if it was a topic that wasn't found in the db? rather than you throwing an error if no articles are found